### PR TITLE
Fix #407

### LINKED ISSUE #407: [2.0-preview] Using arrow keys to navigate causes the the item to be removed
Using arrow keys to focus the next item in the list causes the item to be removed.

### Expected outcome

Using arrow keys to focus the next item should work.

### Actual outcome

The item is removed.

screen shot 2017-03-03 at 5 26 18 pm

### Steps to reproduce

1. Run `demo/cards.html`
2. Scroll down the list.
3. Click on an item to select it.
4. Using `down` arrow key to navigate/focus the next item

### Browsers Affected

- [x] Chrome

### PR DESCRIPTION
**Pull request by blasten**
This bug was introduced because v1 used to rely on the keys provided by `Polymer.Collection` for the mapping between physical items and items. That key was used for things like:  I focused an item, now keep the state when the physical item is off-screen and restore it when it's on the screen.

Now, that is a bit difficult because the items array might mutate causing all kind of issues. This is an attempt to fix that issue _partially_.

### CODE REVIEW

File: iron-list.html [hunk 1]
```diff
@@ -1075,7 +1080,7 @@
           .reduce(function(prev, splice) {
             return prev || this._isIndexRendered(splice.index);
          }.bind(this), false)) {
-         this._debounceRender(this._render, ANIMATION_FRAME);
+          this._debounceRender(this._render, ANIMATION_FRAME);
```
Comment (keanulee) [inline iron-list.html]: Extra space?

File: iron-list.html [hunk 2]
```diff
@@ -1096,11 +1101,9 @@
       if (isIndexRendered) {
         inst = this.modelForElement(this._physicalItems[this._getPhysicalIndex(idx)]);
       } else if (offscreenItem) {
-        // This is only needed for hybrid mode.
-        inst = this.modelForElement(offscreenItem)[this.indexAs] === idx ?
-            this.modelForElement(offscreenItem) : null;
+        inst = this.modelForElement(offscreenItem);
       }
-      if (!inst || (this._collection && inst.__key__ !== key)) {
+      if (!inst || (!IS_V2 && inst.__key__ !== key) || (IS_V2 && inst[this.indexAs] !== idx)) {
```
Comment (keanulee) [inline iron-list.html]: Can V1 (not just V2) use `inst[this.indexAs] !== idx`?

File: iron-list.html [hunk 3]
```diff
@@ -1719,29 +1723,30 @@
         this._itemsParent.appendChild(inst.root);
       }
       // Set the offcreen focused physical item.
-      this._offscreenFocusedItem = this._physicalItems[pidx];
+      this._offscreenFocusedItem = this._physicalItems[fpidx];
       this.modelForElement(this._offscreenFocusedItem).tabIndex = 0;
-      this._physicalItems[pidx] = this._focusBackfillItem;
+      this._physicalItems[fpidx] = this._focusBackfillItem;
+      this._focusedPhysicalIndex = fpidx;
+      //this._focusedItem = this._physicalItems[fpidx];
```
Comment (keanulee) [inline iron-list.html]: commented line

File: iron-list.html [hunk 4]
```diff
@@ -1719,29 +1723,30 @@
...
       // Hide the focused physical.
       this.translate3d(0, HIDDEN_Y, 0, this._offscreenFocusedItem);
     },
 
     _restoreFocusedItem: function() {
-      var pidx, fidx = this._focusedIndex;
-
-      if (!this._offscreenFocusedItem || this._focusedIndex < 0) {
+      var fvidx = this._focusedVirtualIndex;
```
Comment (keanulee) [inline iron-list.html]: Unused variable

File: iron-list.html (hunk 1 above)
Comment (blasten) [inline iron-list.html]: That space fixes the indentation

File: iron-list.html (hunk 2 above)
Comment (blasten) [inline iron-list.html]: I made some changes to simplify the check. so yes, it can just be `inst[this.indexAs] !== idx`.

**Comment by blasten** [comment]
PTAL

File: npm-debug.log [hunk 5]
```diff
@@ -0,0 +1,23 @@
+0 info it worked if it ends with ok
```
Comment (keanulee) [inline npm-debug.log]: Think you checked this in by mistake...

Comment (blasten) [inline npm-debug.log]: oops

**Pull request merged**